### PR TITLE
Shapley Update

### PR DESCRIPTION
**Pull request by Eugene-hu**
Major Changes:

- Updates the scoring method from a fisher information approximation to a mixture of scoring based on partial contribution value and first order gradients. This is balance both short term and long term scoring of peers
- Fixes a bug in peer weights where optimizer stops updating the peer weights. Now we reapply hooks to peer weights after syncing

### CODE REVIEW

File: bittensor/_neuron/text/advanced_server/nucleus_impl.py [hunk 1]
```diff
@@ -221,7 +221,7 @@ def config ():
         parser.add_argument('--neuron.metagraph_sync', type=float, help='how often to sync the metagraph', default=100000)
         parser.add_argument('--neuron.blocks_per_set_weights', type=float, help='how often to sync set weights', default=100)
         parser.add_argument('--neuron.blocks_per_epoch', type=int, help='Blocks per epoch', default=2)
-        parser.add_argument('--neuron.blacklist.time', type=int, help='how often a peer can query you (seconds) ', default=2)
+        parser.add_argument('--neuron.blacklist.time', type=int, help='how often a peer can query you (seconds) ', default=0)
```
Comment (isabella618033) [inline bittensor/_neuron/text/advanced_server/nucleus_impl.py:224]: Maybe later on we should add a 'frequency of request' to our priority function 
as an anti-measure for this change.

File: bittensor/_neuron/text/neuron_utilities.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            jac.append(grad_x.reshape(x.shape)) 
+            pass
+        else:
+            grad_y[i] = 1.
+            grad_x, = torch.autograd.grad(flat_y, x, grad_y, retain_graph=True, create_graph=create_graph)
+            jac.append(grad_x.reshape(x.shape))                                                           
+            grad_y[i] = 0.                                                                                
+    return torch.stack(jac).reshape(y.shape + x.shape)     
+
+def fisher_score_approixmation(loss, peer_weights, ):
```
Comment (shibshib) [inline bittensor/_neuron/text/neuron_utilities.py]: Nit: approximation, not approixmation, typo :)

File: bittensor/_neuron/text/advanced_server/nucleus_impl.py (hunk 1 above)
Comment (Eugene-hu) [reply bittensor/_neuron/text/advanced_server/nucleus_impl.py:224]: Yea, I agree a time component to the priority function makes more sense

File: bittensor/_neuron/text/sgmoe_validator/run.py [hunk 3]
```diff
@@ -156,6 +161,7 @@ def run( config , validator, subtensor, wallet, metagraph, dataset, device, uid,
             ], axis = 1)
             df['uid'] = df.index
             wandb_dendrite = dendrite.to_wandb()
+            #wandb.log(info, step =current_block)
```
Comment (shibshib) [inline bittensor/_neuron/text/sgmoe_validator/run.py]: is this intentional? if so, let's just take out the whole comment.

File: bittensor/_neuron/text/template_miner/__init__.py [hunk 4]
```diff
@@ -92,7 +92,7 @@ def add_args( cls, parser: argparse.ArgumentParser ):
             # ---- Add neuron args.
             parser.add_argument('--config', type=str, help='If set, defaults are overridden by passed file.')
             parser.add_argument('--neuron.learning_rate', type=float, help='Training initial learning rate.', default=1)
-            parser.add_argument('--neuron.learning_rate_chain', type=float, help='Training initial learning rate.', default=1)
+            parser.add_argument('--neuron.learning_rate_chain', type=float, help='Training initial learning rate.', default=0.1)
```
Comment (shibshib) [inline bittensor/_neuron/text/template_miner/__init__.py]: This has the exact same description as the regular learning rate above it. what's the difference?

File: bittensor/_neuron/text/neuron_utilities.py (hunk 2 above)
Comment (shibshib) [reply bittensor/_neuron/text/neuron_utilities.py]: Also, is this function vestigial? I don't see it being used anywhere.

**Review by shibshib — APPROVED** [review]

File: bittensor/_neuron/text/sgmoe_validator/run.py (hunk 3 above)
Comment (isabella618033) [inline bittensor/_neuron/text/sgmoe_validator/run.py]: hahaha

File: bittensor/_neuron/text/template_miner/neuron_impl.py [hunk 5]
```diff
@@ -69,13 +68,16 @@ def __init__( self, config: 'bittensor.config', nucleus: 'Nucleus'):
         self.nucleus.metagraph = self.metagraph_callback
         self.nucleus.dendrite = self.dendrite
         self.optimizer = torch.optim.SGD(
-            [ {'params': self.nucleus.peer_weights, 'lr': self.config.neuron.learning_rate_chain} ],
```
Comment (isabella618033) [inline bittensor/_neuron/text/template_miner/neuron_impl.py:72]: if the learning_rate_chain is not needed here, might as well remove it from the config

File: bittensor/_neuron/text/neuron_utilities.py (hunk 2 above)
Comment (Eugene-hu) [reply bittensor/_neuron/text/neuron_utilities.py]: Haha nice catch, yes, it is vestigial. I left it incase we wish to go back to fisher approximation in the future.

File: bittensor/_neuron/text/sgmoe_validator/run.py (hunk 3 above)
Comment (Eugene-hu) [reply bittensor/_neuron/text/sgmoe_validator/run.py]: Oops, let me remove that

File: bittensor/_neuron/text/template_miner/__init__.py (hunk 4 above)
Comment (Eugene-hu) [reply bittensor/_neuron/text/template_miner/__init__.py]: We don't actually use this anymore, let me remove it.

**Review by isabella618033 — APPROVED** [review]

**Review by unconst — APPROVED** [review]
